### PR TITLE
add error comparator features

### PR DESCRIPTION
**Pull request by ladydascalie**

### CODE REVIEW

File: test/comparators.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+// ErrorTypeComparator defines an error type comparator for use with go-cmp
```
Comment (zeevallin) [inline test/comparators.go]: I'd put this in a `ExampleErrorTypeComparator` clause in the test suite so that it's vetted on build.

File: test/comparators_test.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+type TestError struct{ error }
+
+func TestErrorTypeComparator(t *testing.T) {
+	testErr := TestError{errors.New("oops")}
+	someErr := errors.New("ouch")
+	opts := cmp.Options{
+		test.ErrorTypeComparator,
+	}
+	if cmp.Equal(testErr, someErr, opts) {
+		t.Fatalf("exected unequality between %T and %T\n", someErr, testErr)
```
Comment (zeevallin) [inline test/comparators_test.go]: This seems like some really obtuse language. `exected` is a strange choice of word. Did you mean expected? If so, I would suggest saying `expected %T and %T to be of the same type`.

File: test/comparators.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+// Example:
+//  opts := cmp.Options{
+//  	 test.ErrorTypeComparator,
+//  }
+//
+//  if !cmp.Equal(err, MyCustomError{}, opts) {
+//  	t.Fatal(e.String())
+//  }
+//
+var ErrorTypeComparator = cmp.Comparer(func(x error, y error) bool {
```
Comment (zeevallin) [inline test/comparators.go]: I suggest naming this `ErrorTypeComparer` for consistency and somewhat more correct language.

**Review by zeevallin — CHANGES_REQUESTED** [review]
I would love for this to include some documentation for use in the README.

File: test/comparators_test.go (hunk 2 above)
Comment (ladydascalie) [reply test/comparators_test.go]: I did mean `expected`. I've corrected this.
The point here though is to ensure both types are **not** the same.
I've also corrected that wording.

File: test/comparators.go (hunk 3 above)
Comment (ladydascalie) [reply test/comparators.go]: Good suggestion. Done.

File: test/comparators.go (hunk 1 above)
Comment (ladydascalie) [reply test/comparators.go]: there __is__ a `ExampleErrorTypeComparator` clause in the test file.

**Review by ladydascalie — COMMENTED** [review]
Changes done as per requested

File: test/comparators_test.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/LUSHDigital/core/test"
+	"github.com/google/go-cmp/cmp"
+)
+
+type TestError struct{ error }
+
+func TestErrorTypeComparator(t *testing.T) {
```
Comment (zeevallin) [inline test/comparators_test.go:14]: Should be renamed

File: test/README.md [hunk 5]
```diff
@@ -17,4 +17,34 @@ t.Run("foo equals foo", func(t *testing.T) {
...
+	someErr := errors.New("ouch")
+
+	var e test.ErrorReporter
+	opts := cmp.Options{
+		test.ErrorTypeComparer,
+		cmp.Reporter(&e),
+	}
+	if !cmp.Equal(testErr, someErr, opts) {
+		fmt.Println(e.String())
+	}
```
Comment (zeevallin) [inline test/README.md:43]: Honestly, this feels like quite a lot of boilerplate. I'm wondering if we should provide functions that gave us the error reporter and options as a tuple.

```go
r, opts := test.ReportErrors()
if !cmp.Equal(expected, err, opts) {
    fmt.Println(r.String())
}

```

File: test/comparators_test.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	testErr := TestError{errors.New("oops")}
+	someErr := errors.New("ouch")
+	opts := cmp.Options{
+		test.ErrorTypeComparer,
+	}
+	if cmp.Equal(testErr, someErr, opts) {
+		t.Fatalf("expected %T and %T to be of different types\n", someErr, testErr)
+	}
+}
```
Comment (zeevallin) [inline test/comparators_test.go:23]: I honestly think we should have more test cases for when the values are nil, etc.

**Review by zeevallin — CHANGES_REQUESTED** [review]

**Comment by zeevallin** [comment]
Good work so far. I want to take it slow on this one to make sure we design an API we're happy with.

**Comment by ladydascalie** [comment]
Closing as we decided offline that this would be best suited in the `x` namespace to avoid polluting the core API until a package is deemed ready and solid.
